### DISCUSSION TITLE
reverse engineering of Ingka api

### DISCUSSION THREAD

**Discussion by Ephigenia** [Ephigenia/ikea-availability-checker#36] [Ideas]
work in progress for reverse engineering response from ingka API which seems to return details for cashCarry and clickCollect of a product in all countries stores.

Example Request:

```
curl 'https://api.ingka.ikea.com/cia/availabilities/ru/de?itemNos=69022537&expand=StoresList,Restocks' \
  	-H 'X-Client-ID: b6c117e5-ae61-4ef5-b4cc-e0b1e37f0631' \
	--compressed;
```

The API replies as soon as the `x-client-id` header is present.

**Comment by Ephigenia** [comment]
Example response:

```json
{
  "availabilities": [
    {
      "availableForCashCarry": true,
      "availableForClickCollect": true,
      "buyingOption": {
        "cashCarry": {
          "availability": {
            "probability": {
              "thisDay": {
                "colour": {
                  "rgbDec": "0,255,0",
                  "rgbHex": "00FF00"
                },
                "messageType": "HIGH_IN_STOCK"
              },
              "updateDateTime": "2021-02-22T05:43:37.085Z"
            },
            "quantity": 14,
            "restocks": [
              {
                "earliestDate": "2021-02-23",
                "latestDate": "2021-02-23",
                "quantity": 31,
                "reliability": "HIGH",
                "type": "INTERNAL",
                "updateDateTime": "2021-02-22T05:43:37.085Z"
              }
            ],
            "updateDateTime": "2021-02-22T05:43:37.085Z"
          },
          "range": {
            "inRange": true
          },
          "unitOfMeasure": "PIECE"
        },
        "clickCollect": {
          "range": {
            "inRange": true
          }
        },
        "homeDelivery": {
          "range": {
            "inRange": true
          }
        }
      },
      "classUnitKey": {
        "classUnitCode": "063",
        "classUnitType": "STO"
      },
      "itemKey": {
        "itemNo": "20295930",
        "itemType": "ART"
      }
    },
]
}
```

Unclear how to know which array item belongs to which store. 

Maybe the `classUnitKey` conains the `buCode` of the store, example could be Berlin-Lichtenberg:

```json
{
      "classUnitKey": {
        "classUnitCode": "324",
        "classUnitType": "STO"
      }
}
```

**Reply by vrslev** [reply]
classUnitCode _is_ buCode. If I remember correctly you can get stores around the country with Ingka API

**Reply by Ephigenia** [reply]
Well that would be no problem as there’s a static list of buCode to store names & locations in the source.

**Comment by vrslev** [comment]
In Russia we have two warehouses for online shopping. And this API doesn't provide ￼remaining quantity. It shows only if item is sellable online. 

Even if it were showing quantity it still would be not good enough. 
1. IKEA combines multiple warehouses/stores so more items are available for delivery
2. According to delivery type (to home or pick-up point) different item are not available. I really don't know why.

**Reply by vrslev** [reply]
Currently I work on reverse engineering part of this API which are responsible for authentication (getting guest Bearer token), adding items with specific quantity and checking availability (like you do it in cart during making an order)

**Reply by Ephigenia** [reply]
sounds promising!

**Comment by vrslev** [comment]
Just published new repository with reverse engineering part of IKEA's cart API. [Check it out](https://github.com/vrslev/ikea-api-client)

**Reply by Ephigenia** [reply]
Nice, I’ll need some time to go through the code. I’m not familiar with python but I guess I can manage it. Maybe I can replay the requests in postman first for easier debugging and then try to implement this here. Do you have some real-world examples of how to use the script(s)?

**Reply by vrslev** [reply]
In future I'll add docs with descriptions, endpoint and required headers.

After a while I got sick of manually checking real delivery availability and unavailable items. As I was saying I couldn't find any pattern. Cart API kinda solves my problem

**Comment by Ephigenia** [comment]
I created an issue for this for keeping tracking of this "feature request". Feel free to vote on it to make it more important: https://github.com/Ephigenia/ikea-availability-checker/issues/43

**Comment by gwisp2** [comment]
I noted that IOWS API, that is currently used, doesn't work for some items in Russia: [example item](https://www.ikea.com/ru/ru/p/plugghaest-plugghest-podvesnoy-karman-bezhevyy-60482029/).
`ikea-availability stock --country ru 60482029` crashes with UnhandledPromiseRejection because IOWS API returns 410 GONE.
Russian IKEA website uses INGKA API to check availability.

**Reply by Ephigenia** [reply]
thanks four your feedback, i’m going to check this and see what could be a result

**Reply by Ephigenia** [reply]
@gwisp2 do you have any further information about the INGKA API or differences, anything you know could be helpful!

**Comment by vrslev** [comment]
@Ephigenia I added examples of package usage, feel free to try it out: https://github.com/vrslev/ikea-api-client. Also, I can help to implement this new API in this project if you're willing to implement it.

**Reply by Ephigenia** [reply]
Thank you! I will try to have look this week. It would be great to have help here but maybe your project is having a different scope and it’s better to see it as a kind of migration to nodejs. I will have to understand what’s the difference and which parts are common.

**Reply by Ephigenia** [reply]
> managing cart, checking delivery options, fetching product information and other

that’s a wide list of new and great features
